config/v1/types_proxy.go

```diff
@@ -34,6 +34,14 @@ type ProxySpec struct {
 	// Empty means unset and will not result in an env var.
 	// +optional
 	NoProxy string `json:"noProxy,omitempty"`
+
+	// trustedCA is a reference to a ConfigMap containing a CA certificate bundle used
+	// for client egress HTTPS connections. The certificate bundle must be from the CA
+	// that signed the proxy's certificate and be signed for everything. If trustedCA is
+	// present, consumers are responsible for reading ConfigMapNameReference, copying the ConfigMap
+	// content and using the CA as needed. The namespace for this ConfigMap is "openshift-config".
+	// +optional
+	TrustedCA ConfigMapNameReference `json:"trustedCA,omitempty"`
```

Comment:
this was my proposal but i'd like to see @derekwaynecarr or another architect second it before we make it official.

also this presumably means the installer has to be able to collect CAs from the user and populate this field+configmap during install.

Reply:
@abhinavdahiya what are your thoughts on the installer requirements @bparees highlights above?

Reply:
the configmap is useful when it can be used as a volume to a pod. like the cloudproviderconfig configmap..

For most consumers of this api, that's not possible.. and most of them will need to fetch (latest) proxy object (latest) configmap and then make the request using proxy..
this also means that configmap contents can change and clients need to be reacting to that too..

so personally i think since CAs are not secret, the bundle in the object itself is more ergonomic, but depends if @deads2k and other approvers agree if that's a good API design.

Reply:
The value in making it a configmap reference is that we have numerous places in our cluster config resources where we need a list of CAs.  Making it a reference to a configmap allows an admin to share those CAs if they so choose.  Requiring the CAs be part of each resource (proxy config, build config, image config, etc) requires duplicating the values.

Reply:
in anycase @abhinavdahiya i think the real question to you is, do you foresee issues w/ the installer collecting CAs and populating additional cluster config w/ the CA values (in addition to the proxy config the installer was already going to be collecting and populating)

Reply:
The api here is fulfill-able by installer. my only concern was the consumers of this api need to keep track of an object that updated independently of this object.

Reply:
The advantage of a fixed configmap would be that we can mount it statically into pods, including operators. A proxy object with reference will need pod spec rewrite logic.

I question the value of making this a configmap reference in the proxy object. If we really need such a indirection, we can also make the proxy controller copy it over from another resource. No need to create a complex API.

Reply:
Another point: we have the configmap reference here in the spec, but not in the status. Also the config map can be edited by the admin without any validation through the controller.

An alternative would be: store the CA here in the spec directly as @abhinavdahiya suggests. Let the proxy controller do its endpoint validation with that CA and then it writes it to openshift-config-manager/http-proxy-ca config map. From there pods can statically mount it.

Reply:
> The advantage of a fixed configmap would be that we can mount it statically into pods, including operators.

don't we need to copy it into every operator's namespace anyway to mount it?  I wasn't aware you can mount a configmap from another namespace.

Anyway this is not the data being mounted as it is not in a directly consumable form.  This data is going to be aggregated along w/ the system CAs, into a new config map in openshift-config-managed that has the structure of /etc/pki.  If there is value in having *that* configmap have a static name, we can do so.

> I question the value of making this a configmap reference in the proxy object.

Because you don't believe we'd have reason to reuse the same CA configmap in multiple places?  We already have APIs that consume CAs via a configmap reference, so if this API does not also do that, we are ensuring people must duplicate their CAs.

Reply:
Instead of a ConfigMap this really needs to be the CA string for bootstrap.  In the bootstrap process the CA will need to be added the host's ca trust before trying to pull the images.

Reply:
> Instead of a ConfigMap this really needs to be the CA string for bootstrap. In the bootstrap process the CA will need to be added the host's ca trust before trying to pull the images.

nevermind

Reply:
> Another point: we have the configmap reference here in the spec, but not in the status. Also the config map can be edited by the admin without any validation through the controller.

Again since there is a controller that will be responsible for taking this "user provided CA configmap" and converting it into the "/etc/pki merged configmap" that pods actually consume, that controller can validate the CA/proxy config before it update the "/etc/pki merged configmap" that lives in openshift-config-managed.

Reply:
I like having a configmap reference. ConfigMaps are easy for cluster admins to build and manage. We use references heavily in the rest of our cluster config resources.  Giving flexibility to a cluster-admin to choose locations and specificity for post-processing is a pattern that has worked effectively for other cluster-admin inputs so far.